File: python/taflow/indicators/wedge_pattern.py
"""Python adapter for native wedge-pattern detection."""

from typing import Any
import numpy as np
from .._native import WedgePattern as _Native
from .._series import as_float64_series


class WedgePattern:
    """Detect converging rising and falling wedges from confirmed pivots.

    A rising pair of trendlines whose lower line is steeper emits ``-1.0``; a
    falling pair whose upper line falls faster emits ``1.0``. Other bars emit
    ``0.0``. Geometry uses high and low from four confirmed 5% swings. This
    canonical class maps to Wickra ``Wedge``.

    Args:
        open: Required opening-price history.
        high: Required high-price history.
        low: Required low-price history.
        close: Required closing-price history.

    Raises:
        ValueError: If OHLC histories have different lengths.
    """

    def __init__(self) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native()

    def append(self, open: float, high: float, low: float, close: float) -> "WedgePattern":
        """Append one OHLC bar and return this adapter."""
        self._state.append(float(open), float(high), float(low), float(close))
        return self

    def extend(self, open: Any, high: Any, low: Any, close: Any) -> "WedgePattern":
        """Append aligned OHLC histories after length validation."""
        arrays = tuple(as_float64_series(item) for item in (open, high, low, close))
        if len({len(item) for item in arrays}) != 1:
            raise ValueError("OHLC inputs must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest signal, or ``None`` before the first bar."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return aligned ``-1``, ``0``, and ``1`` signals."""
        return self._state.compute()

    def reset(self) -> "WedgePattern":
        """Clear confirmed pivots and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-bar count delegated to native state."""
        return len(self._state)

File: crates/taflow-core/src/indicators/wedge_pattern.rs
use super::pattern_swing::{recent_legs, SwingTracker, SWING_THRESHOLD};
use crate::error::TaResult;

/// Rising/falling wedge reversal signal from four confirmed pivots.
#[derive(Debug, Clone)]
pub struct WedgePattern {
    swing: SwingTracker,
    count: usize,
    value: Option<f64>,
}

impl WedgePattern {
    /// Create a detector with Wickra's fixed 5% swing threshold.
    pub fn new() -> TaResult<Self> {
        Ok(Self {
            swing: SwingTracker::new(SWING_THRESHOLD, 4),
            count: 0,
            value: None,
        })
    }
    /// Append one OHLC bar and return `1`, `-1`, or `0`.
    pub fn append(&mut self, _open: f64, high: f64, low: f64, _close: f64) -> Option<f64> {
        self.count += 1;
        self.value = Some(0.0);
        if !self.swing.append(high, low) || self.swing.pivots().len() < 4 {
            return self.value;
        }
        let (old_high, new_high, old_low, new_low) = recent_legs(self.swing.pivots());
        let high_slope = new_high - old_high;
        let low_slope = new_low - old_low;
        self.value = if high_slope > 0.0 && low_slope > high_slope {
            Some(-1.0)
        } else if high_slope < low_slope && low_slope < 0.0 {
            Some(1.0)
        } else {
            Some(0.0)
        };
        self.value
    }
    /// Return the signal for the latest bar.
    pub fn value(&self) -> Option<f64> {
        self.value
    }
    /// Return the number of processed bars.
    pub fn len(&self) -> usize {
        self.count
    }
    /// Return whether no bars were processed.
    pub fn is_empty(&self) -> bool {
        self.count == 0
    }
    /// Clear pivots and latest output.
    pub fn reset(&mut self) {
        self.swing.reset();
        self.count = 0;
        self.value = None;
    }
}
